capsules/extra/src/bmm150.rs

```diff
@@ -0,0 +1,230 @@
+                        buffer[1] = 0x3A_u8;
+
+                        if let Err((error, buffer)) = self.i2c.write(buffer, 2) {
+                            self.buffer.replace(buffer);
+                            self.i2c.disable();
+                            self.ninedof_client
+                                .map(|client| client.callback(error as usize, 0, 0));
+                        } else {
+                            self.state.set(State::InitializeReading);
+                        }
+                    }
+                    _ => {}
+                }
+            })
+            .ok_or(ErrorCode::FAIL)
+    }
+}
+
+impl<'a, I: I2CDevice> NineDof<'a> for BMM150<'a, I> {
+    fn set_client(&self, client: &'a dyn NineDofClient) {
+        self.ninedof_client.set(client);
+    }
+
+    fn read_magnetometer(&self) -> Result<(), ErrorCode> {
+        match self.state.get() {
+            State::Suspend => {
+                self.pending_data.set(true);
+                self.initialize()
+            }
+            _ => {
+                if !self.pending_data.get() {
+                    self.pending_data.set(true);
+                    self.start_measurement()
+                } else {
+                    Err(ErrorCode::BUSY)
+                }
+            }
+        }
+    }
+}
+
+#[derive(Clone, Copy, Debug)]
+enum State {
+    Suspend,
+    Sleep,
+    InitializeReading,
+    ReadMeasurement,
+    Read,
+}
+
+impl<'a, I: I2CDevice> I2CClient for BMM150<'a, I> {
+    fn command_complete(&self, buffer: &'static mut [u8], status: Result<(), i2c::Error>) {
+        if let Err(i2c_err) = status {
+            self.state.set(State::Sleep);
+            self.buffer.replace(buffer);
+            self.ninedof_client
+                .map(|client| client.callback(i2c_err as usize, 0, 0));
+            return;
+        }
+
+        match self.state.get() {
+            State::InitializeReading => {
+                buffer[0] = Registers::DATAxLsb as u8;
+
+                if let Err((i2c_err, buffer)) = self.i2c.write(buffer, 1) {
+                    self.state.set(State::Sleep);
+                    self.buffer.replace(buffer);
+                    self.ninedof_client
+                        .map(|client| client.callback(i2c_err as usize, 0, 0));
+                } else {
+                    self.state.set(State::ReadMeasurement);
+                }
+            }
+            State::ReadMeasurement => {
+                if let Err((i2c_err, buffer)) = self.i2c.read(buffer, 8) {
+                    self.state.set(State::Sleep);
+                    self.buffer.replace(buffer);
+                    self.ninedof_client
+                        .map(|client| client.callback(i2c_err as usize, 0, 0));
+                } else {
+                    self.state.set(State::Read);
+                }
+            }
+            State::Read => {
+                let x_axis = ((buffer[1] as i16) << 5) | buffer[0] as i16;
```

Comment:
Is this right?
<img width="1191" alt="image" src="https://github.com/tock/tock/assets/1467890/224153e8-f9b7-4929-b0d0-ba3c36c1c560">
If the resulting value is supposed to be left-aligned (which seems to be what the bmm150 is expecting), then the upper byte needs to be shifted 8. If the bits should be right aligned then the resulting i16 needs to be shifted right 3.

Reply:
I think it's right as is. It specifies exactly which bits it the MSB and LSB occupy. The data sheet says that the MSB should be in bits 12-5 and the LSB in 4-0. The MSB starts at 8-0, so I shift it left by 5 to put it in 12-5. Then I do a logical OR operation which puts the LSB in 4-0.

Reply:
> and the LSB in 4:0

I don't think this is quite right.

<img width="871" alt="image" src="https://github.com/tock/tock/assets/1467890/f9b420a9-6cda-4f49-bae9-3f543835f8d7">

The LSB bits that semantically are 4:0 are stored in 7:3.

Reply:
Your right, I see it now.